backends/dpdk/dpdkAsmOpt.cpp

```diff
@@ -280,5 +264,104 @@ bool ValidateTableKeys::preorder(const IR::DpdkAsmProgram *p) {
+    auto args = new IR::Vector<IR::Argument>();
+    args->push_back(new IR::Argument(new IR::Constant(IR::Type::Bits::get(32), 1)));
+    auto annot = IR::Annotations::empty;
+    annot->addAnnotationIfNew(IR::Annotation::nameAnnotation,
+                              new IR::StringLiteral(instanceName));
+    auto decl = new IR::DpdkExternDeclaration(instanceName, annot, spectype, args,
+                    nullptr);
+    return decl;
+}
+
+void DirectionToRegRead::addMetadataField(cstring fieldName) {
+    if (newFieldName.count(fieldName))
+        return;
+     newMetadataFields.push_back(new IR::StructField(IR::ID(fieldName),
+                                 IR::Type::Bits::get(32)));
+     newFieldName.insert(fieldName);
+}
+
+bool DirectionToRegRead::isDirection(const IR::Member *m) {
+    if (m == nullptr)
+        return false;
+     return m->member.name == "pna_main_input_metadata_direction"
+         || m->member.name == "pna_pre_input_metadata_direction"
+         || m->member.name == "pna_main_parser_input_metadata_direction";
+}
+
+const IR::Node *DirectionToRegRead::postorder(IR::DpdkListStatement *l) {
+    l->statements = replaceDirectionWithRegRead(l->statements);
+    newStmts.clear();
+    return l;
+}
+
+void DirectionToRegRead::replaceDirection(const IR::Member *m) {
+    addMetadataField(reg_read_tmp);
+    addMetadataField(left_shift_tmp);
+    auto reade = new IR::Member(new IR::PathExpression(IR::ID("m")),
+                                IR::ID(reg_read_tmp));
+    auto reads = new IR::DpdkRegisterReadStatement(reade, registerInstanceName,
+                                                   new IR::Constant(IR::Type::Bits::get(32),
+                                                   0));
+    auto shld = new IR::Member(new IR::PathExpression(IR::ID("m")),
+                               IR::ID(left_shift_tmp));
+    auto mov = new IR::DpdkMovStatement(shld, new IR::Constant(IR::Type::Bits::get(32), 1));
+    auto shl = new IR::DpdkShlStatement(shld, shld,
+               new IR::Member(new IR::PathExpression(IR::ID("m")),
+                              IR::ID(dirToInput[m->member.name])));
+    auto mov1 = new IR::DpdkMovStatement(m, reade);
+    auto and0 = new IR::DpdkAndStatement(m, m, shld);
+    newStmts.push_back(reads);
+    newStmts.push_back(mov);
+    newStmts.push_back(shl);
+    newStmts.push_back(mov1);
+    newStmts.push_back(and0);
+}
+
+IR::IndexedVector<IR::DpdkAsmStatement>
+DirectionToRegRead::replaceDirectionWithRegRead(IR::IndexedVector<IR::DpdkAsmStatement> stmts) {
+    for (auto s : stmts) {
+         if (auto jc = s->to<IR::DpdkJmpCondStatement>()) {
+             if (isDirection(jc->src1->to<IR::Member>())) {
+                 replaceDirection(jc->src1->to<IR::Member>());
+             } else if (isDirection(jc->src2->to<IR::Member>())) {
+                    replaceDirection(jc->src2->to<IR::Member>());
+             }
+         } else if (auto u = s->to<IR::DpdkUnaryStatement>()) {
```

Comment:
I think enums only allow equality/inequality comparisons. Do you have a test hitting this block of code?

Reply:
what about meta.dir = istd.direction ?

Reply:
OK. That makes me think of other scenarios such as passing this direction field as parameter to actions/functions., cast operations (which might be simplified at this point). Please verify with such cases once.